herd/AArch64Sem.ml

```diff
@@ -216,44 +221,50 @@ module Make
           (A.Location_reg (ii.A.proc,r)) ii
 
       let neon_setlane old_val idx esize v =
-        let mask = V.op1 (Op.LeftShift (idx*esize)) (AArch64.neon_mask esize) in
+        let mask =
+          V.op1 (Op.LeftShift (idx*esize)) (AArch64.neon_mask esize) in
         let invert = V.op1 Op.Inv mask in
         M.op1 (Op.LeftShift (idx*esize)) v >>= fun new_val ->
         M.op Op.And invert old_val >>|
         M.op Op.And mask new_val >>= fun (v1,v2) ->
         M.op Op.Or v1 v2
 
-      let rec neon_replicate old_val nelem esize v = match nelem with
-      | 0 -> M.unitT old_val
-      | _ ->
-        neon_setlane old_val (nelem-1) esize v >>= fun old_val ->
-        neon_replicate old_val (nelem-1) esize v
+      let rec neon_replicate old_val nelem esize v =
+          if nelem <= 0 then
+            M.unitT old_val
+          else
+            neon_setlane old_val (nelem-1) esize v >>= fun old_val ->
+            neon_replicate old_val (nelem-1) esize v
 
       let write_reg_neon_sz sz r v ii =
         let vr = match r with
         | AArch64Base.SIMDreg _ -> r
         | AArch64Base.Vreg(vr',_) -> (AArch64Base.SIMDreg vr')
         | _ -> assert false in
           (* Clear unused register bits (zero extend) *)
-          uxt_op sz v >>= fun v ->
+          promote v >>= uxt_op sz >>= fun v ->
           let location = A.Location_reg (ii.A.proc,vr) in
           M.write_loc (mk_write MachSize.S128 Annot.N aexp Access.REG v) location ii
 
       let write_reg_neon = write_reg_neon_sz MachSize.S128
 
       let write_reg_neon_elem sz r idx v ii = match r with
       | AArch64Base.Vreg (_,(_,esize)) ->
-          read_reg_neon false r ii >>=
-          fun old_val -> neon_setlane old_val idx esize v >>= fun new_val ->
-          write_reg_neon_sz sz r new_val ii
+         M.data_input_next
```

Comment:
I should not have introduced this change in this very PR. I'll revert.

Reply:
More precisely, consider the following test:
```
AArch64 T
(* Tests MOV from general *)
{
0:X3=16;
}

 P0                     ;
 MOV V0.S[0],W3         ;
locations [0:V0.4S;]
```
The original (and current code after revert) code produces this "sequential" diagram [Seq.pdf](https://github.com/user-attachments/files/15903863/Seq.pdf), while the change produces this "parallel" diagram  [Parallel.pdf](https://github.com/user-attachments/files/15903839/Parallel.pdf).

I think and you may agree that the "parallel" diagram is more intuitive as there is no reason for the the load of the vector register `V0` to depend on the load of the general register `0:X3`. However using the advanced monad combinator `data_input_next` is an overkill. It would be much clearer to use the more standard parallel combinator `>>|` as in:
```
(m1 >>| m2)  >>= fun (v1,v2) ->
```
Where m1 and m2 are the approriate monads (typically `read_reg_ord...` and `read_reg_neon...`.

PS> Monad combinators are  briefly documented in file `herd/monad.mli`. For instance, here are the descriptions of the "sequential" combinators":
```OCaml
 (* Data composition, entry for snd monad: minimals for iico_data *)
    val (>>=) : 'a t -> ('a -> 'b t) -> 'b t
 (* Input to second arg *)
    val data_input_next : 'a t -> ('a -> 'b t) -> 'b t
 (* Input to both args *)
    val data_input_union : 'a t -> ('a -> 'b t) -> 'b t
```